### PR TITLE
VSM improvements and fixes

### PR DESCRIPTION
**Pull request by pixelflinger**
- use EVSM (exponential VSM), helps a lot with light bleeding
- don't apply shadow biases with VSM
- improve VSM user settings: added more controls, expose to java
- minor code cleanups:
   - rename some structures
   - remove unused parameters
   - remove unused uniforms

### CODE REVIEW

**Comment by pixelflinger** [comment]
I marked the PR as draft because I think I need to rework the `VsmShadowOptions` so the variance/light-bleed parameters are per-light.

**Comment by pixelflinger** [comment]
> I marked the PR as draft because I think I need to rework the `VsmShadowOptions` so the variance/light-bleed parameters are per-light.

It's actually pretty involved to move those settings per-light, so if we do it, it'll be in a separate PR.

**Comment by romainguy** [comment]
Can we even move those settings per light? We would have to store more data in the froxels (another case were deferred rendering shines :(). We would also want to keep a global set of options to apply by default to all the lights.

**Comment by pixelflinger** [comment]
> Can we even move those settings per light? We would have to store more data in the froxels (another case were deferred rendering shines :(). We would also want to keep a global set of options to apply by default to all the lights.

I don't think so, it's just a couple of uniforms for the the light-bleed and min variance. I'm not suggesting to be able to chose PCF/VSM per light.

We can't  move mipmaping/anisotropy per-light either because we only have one sampler for all the shadow maps.

**Review by romainguy — COMMENTED** [review]
That looks good but our API names and comments are a bit confusing because we mention both VSM and EVSM. We should make it clear that VSM means EVSM in our case.

**Comment by romainguy** [comment]
> I don't think so, it's just a couple of uniforms for the the light-bleed and min variance. I'm not suggesting to be able to chose PCF/VSM per light.

It's two uniforms *per light* though, so they have to go in the froxels. I know we're not talking about choosing PCF/VSM per light.

**Comment by pixelflinger** [comment]
> > I don't think so, it's just a couple of uniforms for the the light-bleed and min variance. I'm not suggesting to be able to chose PCF/VSM per light.
> 
> It's two uniforms _per light_ though, so they have to go in the froxels. I know we're not talking about choosing PCF/VSM per light.

It's not needed per-froxel, these values are not different per froxel. But in fact we don't even need them per-light, because we only support a limited number of shadows -- so we just need these 2 uniforms per shadowmap, and we already have a UBO per spotlight shadow, so there is a precedent for doing this.

**Comment by romainguy** [comment]
Yes, sorry I meant "light record" not "froxel". But yeah if it's only for the shadow maps that's easier.

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ A new header is inserted each time a *tag* is created.
 - engine: `Backend::DEFAULT` now selects the most appropriate backend for the platform, rather than
   always `OPENGL`. On Android the default is `OPENGL`, on Apple platforms the default is `METAL` and
   on all other platforms that default is `VULKAN`.
+- engine: Use exponential VSM and improve VSM user settings.
```
Comment (bejado) [inline RELEASE_NOTES.md:16]: Make a note about needing to rebuild materials for EVSM?

File: android/filament-android/src/main/java/com/google/android/filament/View.java [hunk 2]
```diff
@@ -701,6 +702,26 @@
          * <strong>Warning: This API is still experimental and subject to change.</strong>
```
Comment (bejado) [inline android/filament-android/src/main/java/com/google/android/filament/View.java:702]: Move this Warning to `VsmShadowOptions` like you did for C++.

File: android/filament-android/src/main/java/com/google/android/filament/View.java [hunk 3]
```diff
@@ -701,6 +702,26 @@
...
          */
         public int anisotropy = 0;
+
+        /**
+         * Whether to generate mipmaps for all VSM shadow maps.
+         */
+        public boolean mipmapping = false;
+
+        /**
+         * EVSM exponent
```
Comment (bejado) [inline android/filament-android/src/main/java/com/google/android/filament/View.java:712]: Copy over C++ documentation for exponent?

File: filament/include/filament/LightManager.h [hunk 4]
```diff
@@ -230,11 +230,13 @@ class UTILS_PUBLIC LightManager : public FilamentAPI {
 
         /** Constant bias in world units (e.g. meters) by which shadows are moved away from the
          * light. 1mm by default.
+         * This is ignored when VSM are used.
```
Comment (bejado) [inline filament/include/filament/LightManager.h:233]: I'd phrase this as:

```
This is ignored when the View's ShadowType is set to VSM.
```

Same for `normalBias`.

File: filament/src/ShadowMap.cpp [hunk 5]
```diff
@@ -90,27 +90,25 @@ void ShadowMap::render(DriverApi& driver, FView::Range const& range, RenderPass&
...
-    const mat4f M = mat4f::lookAt(lightPosition, lightPosition + dir, float3{ 0, 1, 0 });
-    const mat4f Mv = FCamera::rigidTransformInverse(M);
+mat4f ShadowMap::getLightViewMatrix(float3 position, float3 direction) noexcept {
+    const mat4f M = mat4f::lookAt(position, position + direction, float3{ 0, 1, 0 });
+    return FCamera::rigidTransformInverse(M);
+}
+
+void ShadowMap::computeSceneCascadeParams(float3 dir,
+        FScene const& scene, filament::CameraInfo const& camera, uint8_t visibleLayers,
+        SceneInfo& cascadeParams) {
```
Comment (bejado) [inline filament/src/ShadowMap.cpp:100]: Rename `cascadeParams` to `sceneInfo`?

File: filament/src/ShadowMap.cpp [hunk 6]
```diff
@@ -428,32 +431,37 @@ void ShadowMap::computeShadowCameraDirectional(
...
         // It's important to set the light camera's model matrix separately from its projection, so
         // that the cameraPosition uniform gets set correctly.
         // mLightSpace is used in the shader to access the shadow map texture, and has the model
         // matrix baked in.
 
-        mCamera->setModelMatrix(FCamera::rigidTransformInverse(b) * M);
+        // The model matrix below is in fact inverted to get the view matrix and passed to the
+        // shader as 'viewFromWorldMatrix', and is used in the VSM case to compute the depth metric.
+        // (see depth_main.fs). Note that in the case of VSM, 'b' below is identity.
+        mCamera->setModelMatrix(inverse(Mv * b));
```
Comment (bejado) [inline filament/src/ShadowMap.cpp:454]: Can we not use `rigidTransformInverse` here?

File: filament/src/ShadowMap.cpp [hunk 7]
```diff
@@ -624,22 +636,22 @@ mat4f ShadowMap::getTextureCoordsMapping() const noexcept {
...
+    // (the normalization factor is therefore -1/zfar).
+    //
     // When sampling a VSM shadow map, the shader follows suit, and doesn't divide by w for the z
     // coordinate. See getters.fs.
-    math::mat4f lightSpaceVsm = lightSpacePcf;
-    lightSpaceVsm[0][2] = Mv[0][2] * (-1.0f / std::abs(zfar));
-    lightSpaceVsm[1][2] = Mv[1][2] * (-1.0f / std::abs(zfar));
-    lightSpaceVsm[2][2] = Mv[2][2] * (-1.0f / std::abs(zfar));
-    lightSpaceVsm[3][2] = Mv[3][2] * (-1.0f / std::abs(zfar));
+    math::mat4f lightSpaceVsm{ lightSpacePcf };
```
Comment (bejado) [inline filament/src/ShadowMap.cpp:650]: I think we can remove the negative sign here, and also in the shader.

**Review by bejado — COMMENTED** [review]
Overall LGTM. I'll make another pass through later

File: filament/src/ShadowMap.cpp (hunk 5 above)
Comment (pixelflinger) [reply filament/src/ShadowMap.cpp:100]: I realized later that this is both a scene info AND a cascade parameter :-)

File: filament/src/ShadowMap.cpp (hunk 6 above)
Comment (pixelflinger) [reply filament/src/ShadowMap.cpp:454]: Yes, good point.

**Comment by pixelflinger** [comment]
I added a CL that fixes the "receivers are also casters" with VSM in a more robust way.

**Review by romainguy — APPROVED** [review]

File: filament/src/ShadowMap.cpp (hunk 7 above)
Comment (pixelflinger) [reply filament/src/ShadowMap.cpp:650]: I think it'll be less confusing to leave it this way, so we get values between 0 and 1.

Comment (bejado) [reply filament/src/ShadowMap.cpp:650]: Yeah, nevermind I was looking at the wrong equation

**Pull request merged**